File: app/services/supabase_service.py
from supabase import create_client, Client
from app.config import get_settings


class SupabaseService:
    def __init__(self):
        settings = get_settings()
        if settings.supabase_url and settings.supabase_key:
            self.client: Client = create_client(
                settings.supabase_url,
                settings.supabase_key
            )
        else:
            self.client = None

    def is_configured(self) -> bool:
        """Check if Supabase is properly configured."""
        return self.client is not None

    async def get_user_credits(self, user_id: str) -> dict:
        """Get user's remaining credits from Supabase."""
        if not self.is_configured():
            # Return default credits if Supabase not configured
            return {"text": 150, "image": 25, "video": 10, "total": 185}

        try:
            response = self.client.table("user_credits").select(
                "*").eq("user_id", user_id).single().execute()
            if response.data:
                return response.data
            else:
                # Create default credits for new user
                default_credits = {
                    "user_id": user_id,
                    "text": 150,
                    "image": 25,
                    "video": 10
                }
                self.client.table("user_credits").insert(
                    default_credits).execute()
                return {**default_credits, "total": 185}
        except Exception as e:
            print(f"Supabase error: {e}")
            return {"text": 150, "image": 25, "video": 10, "total": 185}

    async def deduct_credit(self, user_id: str, credit_type: str) -> bool:
        """Deduct a credit from user's balance."""
        if not self.is_configured():
            return True  # Skip credit deduction if not configured

        try:
            # Get current credits
            response = self.client.table("user_credits").select(
                credit_type).eq("user_id", user_id).single().execute()

            if response.data and response.data.get(credit_type, 0) > 0:
                new_value = response.data[credit_type] - 1
                self.client.table("user_credits").update(
                    {credit_type: new_value}).eq("user_id", user_id).execute()
                return True
            return False
        except Exception as e:
            print(f"Supabase error: {e}")
            return True  # Allow operation to continue

    async def save_generated_content(self, user_id: str, content: dict) -> bool:
        """Save generated content to history."""
        if not self.is_configured():
            return True

        try:
            self.client.table("content_history").insert({
                "user_id": user_id,
                "platform": content.get("platform"),
                "content_type": content.get("content_type"),
                "caption": content.get("caption"),
                "hashtags": content.get("hashtags"),
                "cta": content.get("cta")
            }).execute()
            return True
        except Exception as e:
            print(f"Supabase error: {e}")
            return False


# Singleton instance
_supabase_service = None


def get_supabase_service() -> SupabaseService:
    global _supabase_service
    if _supabase_service is None:
        _supabase_service = SupabaseService()
    return _supabase_service

File: services/supabaseService.js
import { createClient } from '@supabase/supabase-js';
import { getSettings } from '../config/index.js';

class SupabaseService {
    constructor() {
        const settings = getSettings();
        this.client = null;

        if (settings.supabaseUrl && settings.supabaseKey) {
            try {
                this.client = createClient(settings.supabaseUrl, settings.supabaseKey);
                console.log('✓ Supabase client initialized');
            } catch (e) {
                console.error(`✗ Supabase init failed: ${e.message}`);
            }
        } else {
            console.warn('⚠️  Supabase not configured - database features disabled');
        }
    }

    isConfigured() {
        return this.client !== null;
    }

    async getUserCredits(userId) {
        if (!this.isConfigured()) {
            // Return default credits if Supabase not configured
            return { text: 150, image: 25, video: 10, total: 185 };
        }

        try {
            const { data, error } = await this.client
                .from('user_credits')
                .select('*')
                .eq('user_id', userId)
                .single();

            if (error && error.code !== 'PGRST116') {
                throw error;
            }

            if (data) {
                return data;
            } else {
                // Create default credits for new user
                const defaultCredits = {
                    user_id: userId,
                    text: 150,
                    image: 25,
                    video: 10
                };

                await this.client
                    .from('user_credits')
                    .insert(defaultCredits);

                return { ...defaultCredits, total: 185 };
            }
        } catch (e) {
            console.error(`Supabase error: ${e.message}`);
            return { text: 150, image: 25, video: 10, total: 185 };
        }
    }

    async deductCredit(userId, creditType) {
        if (!this.isConfigured()) {
            return true; // Skip credit deduction if not configured
        }

        try {
            // Get current credits
            const { data, error } = await this.client
                .from('user_credits')
                .select(creditType)
                .eq('user_id', userId)
                .single();

            if (error) throw error;

            if (data && (data[creditType] || 0) > 0) {
                const newValue = data[creditType] - 1;
                await this.client
                    .from('user_credits')
                    .update({ [creditType]: newValue })
                    .eq('user_id', userId);
                return true;
            }
            return false;
        } catch (e) {
            console.error(`Supabase error: ${e.message}`);
            return true; // Allow operation to continue
        }
    }

    async saveGeneratedContent(userId, content) {
        if (!this.isConfigured()) {
            return true;
        }

        try {
            await this.client.from('content_history').insert({
                user_id: userId,
                platform: content.platform,
                content_type: content.contentType,
                caption: content.caption,
                hashtags: content.hashtags,
                cta: content.cta
            });
            return true;
        } catch (e) {
            console.error(`Supabase error: ${e.message}`);
            return false;
        }
    }

    async uploadImageToStorage(imageBase64, filename, bucket = 'ugc-ads') {
        console.log(`📤 Starting image upload process...`);
        console.log(`   Bucket: ${bucket}`);
        console.log(`   Filename: ${filename}`);
        
        if (!this.isConfigured()) {
            throw new Error('Supabase is not configured');
        }

        try {
            console.log(`   Removing data URI prefix...`);
            // Remove data URI prefix if present
            let base64Data = imageBase64;
            if (imageBase64.startsWith('data:')) {
                base64Data = imageBase64.split(',')[1];
            }

            console.log(`   Converting base64 to buffer... (${base64Data.length} characters)`);
            // Decode base64 to bytes
            const imageBuffer = Buffer.from(base64Data, 'base64');
            console.log(`   Buffer created: ${imageBuffer.length} bytes`);

            // Generate unique filename with timestamp
            const timestamp = new Date().toISOString().replace(/[-:]/g, '').split('.')[0];
            const randomId = Math.random().toString(36).substring(2, 10);
            const uniqueFilename = `${timestamp}_${randomId}_${filename}`;
            console.log(`   Unique filename: ${uniqueFilename}`);

            // Upload to Supabase Storage
            console.log(`   Uploading to Supabase storage...`);
            const { data, error } = await this.client.storage
                .from(bucket)
                .upload(uniqueFilename, imageBuffer, {
                    contentType: 'image/png',
                    upsert: false
                });

            if (error) {
                console.error(`   ❌ Upload error: ${JSON.stringify(error)}`);
                throw error;
            }
            
            console.log(`   ✓ Upload successful! Data: ${JSON.stringify(data)}`);

            // Get public URL
            console.log(`   Getting public URL...`);
            const { data: urlData } = this.client.storage
                .from(bucket)
                .getPublicUrl(uniqueFilename);

            console.log(`✅ Image uploaded successfully: ${urlData.publicUrl}`);
            return urlData.publicUrl;

        } catch (e) {
            console.error(`❌ CRITICAL ERROR in uploadImageToStorage:`);
            console.error(`   Message: ${e.message}`);
            console.error(`   Stack: ${e.stack}`);
            // DON'T return base64 - throw the error so we can see what's wrong!
            throw new Error(`Supabase upload failed: ${e.message}`);
        }
    }
}

// Singleton instance
let supabaseService = null;

export function getSupabaseService() {
    if (!supabaseService) {
        supabaseService = new SupabaseService();
    }
    return supabaseService;
}

export default SupabaseService;
